artichoke-backend/src/extn/core/array/trampoline.rs

```diff
@@ -142,6 +142,31 @@ pub fn clear(interp: &mut Artichoke, mut ary: Value) -> Result<Value, Error> {
     Ok(ary)
 }
 
+pub fn push<I>(interp: &mut Artichoke, mut ary: Value, others: I) -> Result<Value, Error>
+where
+    I: IntoIterator<Item = Value>,
+    I::IntoIter: Clone,
+{
+    if ary.is_frozen(interp) {
+        return Err(FrozenError::with_message("can't modify frozen Array").into());
+    }
+
+    let mut array = unsafe { Array::unbox_from_value(&mut ary, interp)? };
+
+    let array_mut = unsafe { array.as_inner_mut() };
```

Comment:
Just to walk through what's happening here in the code you've written:

Grab a mutable reference to the underlying Rust array which we are viewing from a boxed Ruby value. This operation is unsafe because a mutable reference allows us to modify the array and cause a reallocation, which will invalidate the way the Array is stored in the boxed Ruby value and leave a dangling pointer.

Then we push some values into the array. This may cause the array to be reallocated.

The initial use of unsafe is safe though because immediately after a potential reallocation we consume the mutable reference and repack the arrays raw parts into the boxed Ruby value.

The important thing is we don't cause any interpreter heap allocations while the boxed Ruby value is potentially invalidated. If we were to do that, it could trigger a GC which would likely cause a crash when marking the array we are operating on.

Would you be able to add a `SAFETY ` comment similar to the other ones in this file that captures this proof of correctness?

Reply:
Thanks for the detailed explanation. I think I had the idea abstractly but its been a while since I've had to actually think about pointers (I mainly use Ruby in my day-to-day now). 

Just to make sure I have the right understanding of the bad-path: incorrect behavior would be attempting to mutate the `mut array` in a manner that might cause reallocations before the repacking from `as_inner_mut` e.g. calling `array.extend` directly after the first `unsafe`?

Reply:
there would be bad behavior if after the call to `array_mut.extend` but before reboxing `array` into `ary` we called something like `interp.try_convert_mut("abc")` and allocated a Ruby String. This allocation could cause a GC. When the GC visited `ary` to try and mark its children, `ary` would contain an invalid pointer.

For a concrete example and real Artichoke bug due to this, take a peek at:

- https://github.com/artichoke/artichoke/pull/1323